### PR TITLE
Add bit-vector encoding for CTLZ

### PR DESCRIPTION
**Pull request by hernanponcedeleon**
This PR adds a the encoding of the ctlz instruction. I tested it with the following program
```
#include <stdint.h>
#include <assert.h>

volatile int32_t x = 1;
volatile int32_t y = INT32_MAX;
volatile int32_t z = INT32_MAX+1;
volatile int32_t u;

int main()
{
    // x = 0000 0000 0000 0000 0000 0000 0000 0001 
    u = __builtin_clz(x);
    assert(u == 31);

    // y = 0111 1111 1111 1111 1111 1111 1111 1111 
    u = __builtin_clz(y);
    assert(u == 1);

    // z = 1000 0000 0000 0000 0000 0000 0000 0000 
    u = __builtin_clz(z);
    assert(u == 0);

    return 0;
}
```
However, since we are still wrongly parsing types, testing this requires the following flag `--modeling.precision=32`

### CODE REVIEW

File: dartagnan/src/main/java/com/dat3m/dartagnan/expression/op/IOpUn.java [hunk 1]
```diff
@@ -164,11 +163,23 @@ public Formula encode(Formula e, FormulaManager m) {
                     return bvmgr.extend(bv, 48, true);
                 case SEXT3264:
                     return bvmgr.extend(bv, 32, true);
+                case CTLZ:
+                    // enc = extract(bv, 63, 63) == 1 ? 0 : (1 + extract(bv, 62, 62) == 1 ? 0 : 1 + extract ...)
+                    IntegerFormula enc = imgr.makeNumber(0);
+                    for(int i = 0; i < bvmgr.getLength(bv) ; i++) {
+                        enc = ctlzEncodingAtIndex(i, bv, enc, m);
+                    }
+                    return enc;
```
[the remark this answers is not in the document]
Comment (hernanponcedeleon) [reply dartagnan/src/main/java/com/dat3m/dartagnan/expression/op/IOpUn.java]: I guess you meant that if we return in the first ITE, the value is zero
```
ctrl(x) = ITE(extract(x, 63, 63) == 1, 0, ITE(extract(x, 62, 62) == 1, 1, ... ITE(extract(x, 0, 0) == 1, 63, 64))
```

[the remark this answers is not in the document]
Comment (ThomasHaas) [reply dartagnan/src/main/java/com/dat3m/dartagnan/expression/op/IOpUn.java]: Yes, I was off by one :)

**Comment by ThomasHaas** [comment]
I find the code unnecessarily hard to follow. It is awkward that the initial encoding
```
BitvectorFormula enc = bvmgr.makeBitvector(bvmgr.getLength(bv), bvmgr.getLength(bv));
```
is done in one function, and then the remaining encoding is done in a second function, but that second function seems to be strongly coupled with the first one (I don't see how it would make sense if called in isolation).

I think you would get cleaner code if you just inlined the extra function. You would also get rid of most code inside the loop:
```
        BitvectorFormulaManager bvmgr = m.getBitvectorFormulaManager(); // already defined in context of calling function
        int bvLength = bvmgr.getLength(bv); // Can be obtained only once in the calling function (its used there as well)
        if(i == bvLength) { // becomes obsolete if you start the loop at bvLength - 1
            return rest;
        }
        // The only line of code that would remain inside the loop. Creating "bmgr" and "bv1 = bvmgr.makeBitvector(1, 1)" as variables only once also makes this code more readable.
        return m.getBooleanFormulaManager().ifThenElse(bvmgr.equal(bvmgr.extract(bv, bvLength - (i + 1), bvLength - (i + 1)), bvmgr.makeBitvector(1, 1)), bvmgr.makeBitvector(bvLength, i), rest);
```

**Review by ThomasHaas — APPROVED** [review]
LGTM
